RationaleMCP/0031/concept.md

```diff
@@ -0,0 +1,135 @@
+   * packages, [expandable] connectors, models (only one model remains), blocks, operator definitions
+   * functional inputs of functions
+   * conditional declarations
+   * modifications (therefore no merging of them)
+   * connect statements
+   * loops inside equation sections
+   * `if`/`then`/`else` in equation sections (atleast with different sized branches)
+   * `when` clauses
+   * stream operators: `actualStream(...)`, `inStream(...)`
+   * all prefixes, except `input` and `output`
+* __TODO__ synchronous Modelica extensions, including state machines, possibly introduction of partitions
+
+### Names
+
+* Names are defined as the `IDENT` token in Modelica
+* The hierarchy of the Modelica model is encoded into the names of the variables and types by separation of the levels by a dot.
+* Names from Modelica are generally enclosed in single quotation marks, whether or not they contain a dot.  
+* Generated helper variables do not need single quotes. 
+* __TODO__: array components, combining names for helper variables
+
+### Types
+
+* Following types exist:
+   * predefined types: `Real`, `Integer`, `Boolean`, `String`, `ExternalObject`,
+   * record types
+   * functions
+   * simple types.
+   
+* Properties of record definitions are:
+   * The components are defined as variable definitions. Declaration equations are not allowed.
+   * There are no equations or algorithms.
+   * record definitions can not nest.
+   * There is no automatism to generate record constructors anymore. They have to be made explicit as functions.
+
+* Simple types are aliases to predefined types enriched with annotations. 
+   * They are used to avoid duplication of annotations in variable definitions. 
+   * An annotation of a variable definition overrides the same annotation of the type.
+   * They have no dimension.
+
+* The order of definition is not important.
+
+### Variables
+
+* A variable definition consists of
+   1. a prefix (e.g. `input`,`output`),
+   1. a type name,
+   1. a variable name
+   1. an optional dimension specification,
+   1. an optional declaration equations,
+   1. an optional description string and
+   1. an optional annotation.
+* The dimension of an array is defined by a list of dimension sizes inside square brackets, e.g. `[2,3]`. The dimension sizes are allowed to be: 
+   * integer literals,
+   * `:`. In this case is determined from the dimension of the right side of the declaration equation, if exist and if it is possible. If not, the size is seen as unknown and no subscription is possible.
+   * Only in functions: expressions containing literals and references to input parameters and other (predefined) pure functions
+* Parameters and Constants from Modelica are mapped to variables with special annotations.
+* It is assumed, that constants and parameters of the Modelica model was evaluated already evaluated if needed, i.e. because dimensions or branches or loops in equations depend from them.
```

Comment:
Let's make a separate design document for the handling of parameters evaluated during reduction to Flat Modelica.